docs(design): add latency definition, records service and visualization (#106)
tier4/CARET_doc: docs/design/latency_definitions/communication.md

Since ROS communication is capable of many-to-many communication, there are cases where both intra-process and inter-process communication are performed in a single publish.
In CARET, communication is divided into 1:1 pairs and latency is calculated.

## Intra process communication

A simplified sequence diagram focusing only on the relevant data flow is shown below.

```plantuml
@startuml

participant "UserCode\n(Publisher Side)" as Callback
participant "rclcpp\n (Publisher/Subscription)" as Rclcpp
participant "rclcpp\n (Buffer)" as Buffer
participant rcl
participant LTTng

== Publisher ==

-> Callback
activate Callback

Callback -> Rclcpp : publish
activate Rclcpp
Rclcpp -> LTTng : intra process publish

Rclcpp -> Buffer : enqueue
activate Buffer
Rclcpp -> rcl : notify
deactivate Rclcpp
activate rcl

deactivate Callback

== Subscription ==

rcl -> Rclcpp : dispatch
deactivate rcl
activate Rclcpp

Rclcpp -> Buffer : take
Buffer -> Rclcpp: dequeue
deactivate Buffer
Rclcpp -> LTTng: sample dispatch_intra_subscription_callback
Rclcpp -> LTTng : sample callback_start
Rclcpp -> Callback : callback start

activate Callback
@enduml
```

> `to_dataframe` API returns following columns.

Reviewer: ```suggestion
`to_dataframe` API returns a table which has the following columns.
```

Author: addressed

- [RuntimeDataProvider API](https://tier4.github.io/CARET_analyze/latest/infra/#caret_analyze.infra.lttng.lttng.Lttng.compose_intra_proc_comm_records)

## Inter process communication

A simplified sequence diagram focusing only on the relevant data flow is shown below.

```plantuml
@startuml
title: Definition of major tracepoints

participant UserCode
participant ROS2
participant DDS
participant LTTng

== Publisher Side ==

activate ROS2
activate UserCode
UserCode -> ROS2: publish()
activate ROS2
ROS2 -> LTTng: sample rclcpp_publish

ROS2 -> DDS: dds_write()
deactivate ROS2
activate DDS
DDS -> LTTng: sample dds_write
DDS -> LTTng: sample dds_bind_addr_to_stamp

UserCode -> ROS2 : callback_end
deactivate UserCode
deactivate ROS2

== Subscription Side ==

DDS -> ROS2: <notify> on_data_available
activate ROS2

ROS2 -> DDS :  take messages
DDS -> ROS2
deactivate DDS
ROS2 -> LTTng: sample dispatch_subscription_callback
ROS2 -> LTTng: sample [callback_start
ROS2 -> UserCode: callback start
activate UserCode
UserCode -> ROS2: callback end
deactivate UserCode

deactivate ROS2
@enduml
```

> `to_dataframe` API returns following columns.

Reviewer: ```suggestion
`to_dataframe` API returns a table which has the following columns.
```

# Communication

> The communication latency is the time difference between the time the message is passed from the ownership of the callback to the next callback.

Reviewer: ```suggestion
Communication latency is an expression of how much time it takes for a topic message to travel from source callback to next callback.
```

Author: 28579f5

$$

> l_{comm} = t_{receive} - t_{transfer}

Reviewer: ```suggestion
l_{comm} = t_{reception} - t_{transmission}
```

Author: removed: 
28579f5

$$

> In particular, if transfer of ownership is publish and receive of ownership is subscribe, it is defined as follows

Reviewer: As transmission is replaced by publish, but reception is subscription in the context of ROS 2, the expression is translated into the next one.

> In particular, if transfer of ownership is publish and receive of ownership is subscribe, it is defined as follows

Reviewer: ```suggestion
As transmission is replaced by publish and reception is subscription in the context of ROS 2, the equation is translated into the next one.
```

Author: removed: 
28579f5

$$
l_{comm} = t_{sub} - t_{pub}
$$

<prettier-ignore-start>
!!! Info

>         In this definition, communication latency is now affected by scheduling.

Reviewer: ```suggestion
        In this definition, communication latency is now affected by scheduling latency.
```

Author: 28579f5

        It is not only the communication latency of the DDS.

>         If multiple callbacks can be executed at the same time, it may include the execution time of other callbacks.

Reviewer: I cannot understand the intention of this sentence.

Author: removed 
28579f5
